File: arx_svg_parser/services/symbol_renderer.py
import logging
from typing import Dict, List, Any, Optional, Tuple
import xml.etree.ElementTree as ET
from uuid import uuid4
from datetime import datetime
import math

logger = logging.getLogger(__name__)

class SymbolRenderer:
    """Renders recognized symbols into dynamic SVG-BIM elements."""
    
    def __init__(self):
        self.svg_namespace = "http://www.w3.org/2000/svg"
        self.arx_namespace = "http://arxos.io/svg"
        
    def render_recognized_symbols(self, svg_content: str, recognized_symbols: List[Dict[str, Any]], 
                                building_id: str, floor_label: str) -> Dict[str, Any]:
        """
        Render recognized symbols into the SVG-BIM.
        
        Args:
            svg_content: Original SVG content
            recognized_symbols: List of recognized symbols from recognition engine
            building_id: Building identifier
            floor_label: Floor label
            
        Returns:
            Dict containing updated SVG and metadata
        """
        try:
            root = ET.fromstring(svg_content)
            
            # Create or find the arx-objects group
            arx_group = self._get_or_create_arx_group(root)
            
            # Track rendered symbols
            rendered_symbols = []
            
            # Render each recognized symbol
            for symbol_info in recognized_symbols:
                if symbol_info['confidence'] >= 0.5:  # Only render high-confidence matches
                    rendered_symbol = self._render_single_symbol(
                        arx_group, symbol_info, building_id, floor_label
                    )
                    if rendered_symbol:
                        rendered_symbols.append(rendered_symbol)
            
            # Convert back to string
            updated_svg = ET.tostring(root, encoding='unicode')
            
            return {
                'svg': updated_svg,
                'rendered_symbols': rendered_symbols,
                'total_recognized': len(recognized_symbols),
                'total_rendered': len(rendered_symbols),
                'building_id': building_id,
                'floor_label': floor_label
            }
            
        except ET.ParseError as e:
            logger.error(f"Failed to parse SVG for rendering: {e}")
            return {
                'svg': svg_content,
                'error': f"SVG parsing failed: {str(e)}",
                'rendered_symbols': []
            }
    
    def _get_or_create_arx_group(self, root: ET.Element) -> ET.Element:
        """Get or create the arx-objects group in the SVG."""
        # Look for existing arx-objects group
        for group in root.findall(f'.//{{{self.svg_namespace}}}g'):
            if group.get('id') == 'arx-objects':
                return group
        
        # Create new arx-objects group
        arx_group = ET.Element(f'{{{self.svg_namespace}}}g', {
            'id': 'arx-objects',
            'class': 'arx-dynamic-objects'
        })
        root.append(arx_group)
        return arx_group
    
    def _render_single_symbol(self, arx_group: ET.Element, symbol_info: Dict[str, Any], 
                            building_id: str, floor_label: str) -> Optional[Dict[str, Any]]:
        """Render a single recognized symbol."""
        try:
            symbol_id = symbol_info['symbol_id']
            symbol_data = symbol_info['symbol_data']
            confidence = symbol_info['confidence']
            match_type = symbol_info['match_type']
            
            # Generate unique object ID
            object_id = f"{symbol_id}_{uuid4().hex[:8]}"
            
            # Determine position
            position = self._determine_symbol_position(symbol_info, arx_group)
            
            # Create the symbol element
            symbol_element = self._create_symbol_element(
                symbol_id, symbol_data, object_id, position, confidence, match_type
            )
            
            # Add to arx-group
            arx_group.append(symbol_element)
            
            # Create metadata
            metadata = {
                'object_id': object_id,
                'symbol_id': symbol_id,
                'symbol_name': symbol_data.get('display_name', symbol_id),
                'system': symbol_data.get('system', 'unknown'),
                'category': symbol_data.get('category', ''),
                'position': position,
                'confidence': confidence,
                'match_type': match_type,
                'rendered_at': datetime.utcnow().isoformat(),
                'building_id': building_id,
                'floor_label': floor_label,
                'properties': symbol_data.get('properties', []),
                'connections': symbol_data.get('connections', []),
                'tags': symbol_data.get('tags', [])
            }
            
            logger.info(f"Rendered symbol {symbol_id} at position {position}")
            return metadata
            
        except Exception as e:
            logger.error(f"Failed to render symbol {symbol_info.get('symbol_id', 'unknown')}: {e}")
            return None
    
    def _determine_symbol_position(self, symbol_info: Dict[str, Any], arx_group: ET.Element) -> Dict[str, float]:
        """Determine the position for a symbol based on recognition context."""
        position = {'x': 0, 'y': 0}
        
        # If we have a position from SVG text recognition, use it
        if 'position' in symbol_info:
            position = symbol_info['position']
        else:
            # Calculate position based on existing symbols
            position = self._calculate_auto_position(arx_group, symbol_info)
        
        return position
    
    def _calculate_auto_position(self, arx_group: ET.Element, symbol_info: Dict[str, Any]) -> Dict[str, float]:
        """Calculate automatic position for a symbol."""
        # Get existing symbol positions
        existing_positions = []
        for elem in arx_group.findall(f'.//{{{self.svg_namespace}}}g'):
            transform = elem.get('transform', '')
            if 'translate(' in transform:
                # Extract position from transform
                try:
                    # Parse translate(x,y) from transform
                    import re
                    match = re.search(r'translate\(([^,]+),([^)]+)\)', transform)
                    if match:
                        x = float(match.group(1))
                        y = float(match.group(2))
                        existing_positions.append((x, y))
                except (ValueError, AttributeError):
                    pass
        
        # Calculate new position
        if existing_positions:
            # Find a position that doesn't overlap
            base_x, base_y = existing_positions[-1]  # Start from last position
            offset = 50  # 50 units offset
            
            # Try positions in a grid pattern
            for i in range(10):  # Try 10 different positions
                x = base_x + (i % 3) * offset
                y = base_y + (i // 3) * offset
                
                # Check if position is far enough from existing symbols
                if not any(math.sqrt((x-ex_x)**2 + (y-ex_y)**2) < 30 for ex_x, ex_y in existing_positions):
                    return {'x': x, 'y': y}
            
            # If no good position found, just offset from last
            last_x, last_y = existing_positions[-1]
            return {'x': last_x + offset, 'y': last_y + offset}
        else:
            # First symbol - place at origin
            return {'x': 100, 'y': 100}
    
    def _create_symbol_element(self, symbol_id: str, symbol_data: Dict[str, Any], 
                             object_id: str, position: Dict[str, float], 
                             confidence: float, match_type: str) -> ET.Element:
        """Create the SVG element for a symbol."""
        # Get the symbol's SVG content
        symbol_svg = symbol_data.get('svg', '')
        
        # Create the main group element
        group_attrs = {
            'id': object_id,
            'class': f'arx-symbol arx-{symbol_id}',
            'transform': f'translate({position["x"]},{position["y"]})',
            'data-symbol-id': symbol_id,
            'data-symbol-name': symbol_data.get('display_name', symbol_id),
            'data-system': symbol_data.get('system', 'unknown'),
            'data-category': symbol_data.get('category', ''),
            'data-confidence': str(confidence),
            'data-match-type': match_type,
            'data-rendered-at': datetime.utcnow().isoformat()
        }
        
        group = ET.Element(f'{{{self.svg_namespace}}}g', group_attrs)
        
        # Add the symbol SVG content
        if symbol_svg:
            try:
                symbol_root = ET.fromstring(symbol_svg)
                # Copy all child elements
                for child in symbol_root:
                    group.append(child)
            except ET.ParseError:
                # Fallback: add as text
                text_elem = ET.Element(f'{{{self.svg_namespace}}}text', {
                    'x': '0', 'y': '0',
                    'font-size': '12',
                    'text-anchor': 'middle'
                })
                text_elem.text = symbol_data.get('display_name', symbol_id)
                group.append(text_elem)
        
        # Add confidence indicator
        if confidence < 0.8:
            confidence_indicator = ET.Element(f'{{{self.svg_namespace}}}circle', {
                'cx': '0', 'cy': '0', 'r': '15',
                'fill': 'none',
                'stroke': '#ff6b6b' if confidence < 0.6 else '#ffd93d',
                'stroke-width': '2',
                'stroke-dasharray': '5,5',
                'opacity': '0.7'
            })
            group.append(confidence_indicator)
        
        # Add metadata text
        metadata_text = ET.Element(f'{{{self.svg_namespace}}}text', {
            'x': '0', 'y': '25',
            'font-size': '8',
            'text-anchor': 'middle',
            'fill': '#666',
            'class': 'arx-metadata'
        })
        metadata_text.text = f"{symbol_data.get('display_name', symbol_id)} ({confidence:.1f})"
        group.append(metadata_text)
        
        return group
    
    def update_symbol_position(self, svg_content: str, object_id: str, 
                             new_position: Dict[str, float]) -> str:
        """Update the position of an existing symbol."""
        try:
            root = ET.fromstring(svg_content)
            
            # Find the symbol element
            symbol_elem = root.find(f'.//*[@id="{object_id}"]')
            if symbol_elem is not None:
                # Update transform
                symbol_elem.set('transform', f'translate({new_position["x"]},{new_position["y"]})')
                
                # Update metadata
                symbol_elem.set('data-updated-at', datetime.utcnow().isoformat())
            
            return ET.tostring(root, encoding='unicode')
            
        except ET.ParseError as e:
            logger.error(f"Failed to update symbol position: {e}")
            return svg_content
    
    def remove_symbol(self, svg_content: str, object_id: str) -> str:
        """Remove a symbol from the SVG."""
        try:
            root = ET.fromstring(svg_content)
            
            # Find and remove the symbol element
            symbol_elem = root.find(f'.//*[@id="{object_id}"]')
            if symbol_elem is not None:
                # Find parent by searching through all elements
                for parent in root.iter():
                    if symbol_elem in list(parent):
                        parent.remove(symbol_elem)
                        break
            
            return ET.tostring(root, encoding='unicode')
            
        except ET.ParseError as e:
            logger.error(f"Failed to remove symbol: {e}")
            return svg_content
    
    def get_rendered_symbols(self, svg_content: str) -> List[Dict[str, Any]]:
        """Extract metadata from rendered symbols in SVG."""
        symbols = []
        
        try:
            root = ET.fromstring(svg_content)
            
            for group in root.findall(f'.//{{{self.svg_namespace}}}g'):
                if group.get('class', '').startswith('arx-symbol'):
                    symbol_data = {
                        'object_id': group.get('id'),
                        'symbol_id': group.get('data-symbol-id'),
                        'symbol_name': group.get('data-symbol-name'),
                        'system': group.get('data-system'),
                        'category': group.get('data-category'),
                        'confidence': float(group.get('data-confidence', 0)),
                        'match_type': group.get('data-match-type'),
                        'rendered_at': group.get('data-rendered-at')
                    }
                    
                    # Extract position from transform
                    transform = group.get('transform', '')
                    if 'translate(' in transform:
                        import re
                        match = re.search(r'translate\(([^,]+),([^)]+)\)', transform)
                        if match:
                            symbol_data['position'] = {
                                'x': float(match.group(1)),
                                'y': float(match.group(2))
                            }
                    
                    symbols.append(symbol_data)
            
        except ET.ParseError as e:
            logger.error(f"Failed to extract rendered symbols: {e}")
        
        return symbols 

File: arx-backend/services/symbols/symbol_renderer.go
package symbols

import (
	"fmt"
	"math"
	"regexp"
	"strings"
	"time"

	"github.com/google/uuid"
)

// RenderedSymbol represents a rendered symbol in SVG
type RenderedSymbol struct {
	ObjectID    string                 `json:"object_id"`
	SymbolID    string                 `json:"symbol_id"`
	SymbolName  string                 `json:"symbol_name"`
	System      string                 `json:"system"`
	Category    string                 `json:"category"`
	Position    Point                  `json:"position"`
	Confidence  float64                `json:"confidence"`
	MatchType   string                 `json:"match_type"`
	RenderedAt  time.Time              `json:"rendered_at"`
	BuildingID  string                 `json:"building_id"`
	FloorLabel  string                 `json:"floor_label"`
	Properties  map[string]interface{} `json:"properties"`
	Connections []Connection           `json:"connections"`
	Tags        []string               `json:"tags"`
}

// RenderResult represents the result of rendering symbols
type RenderResult struct {
	SVG             string           `json:"svg"`
	RenderedSymbols []RenderedSymbol `json:"rendered_symbols"`
	TotalRecognized int              `json:"total_recognized"`
	TotalRendered   int              `json:"total_rendered"`
	BuildingID      string           `json:"building_id"`
	FloorLabel      string           `json:"floor_label"`
	Error           string           `json:"error,omitempty"`
}

// SymbolRenderer provides rendering functionality for symbols
type SymbolRenderer struct {
	svgNamespace string
	arxNamespace string
}

// NewSymbolRenderer creates a new symbol renderer
func NewSymbolRenderer() *SymbolRenderer {
	return &SymbolRenderer{
		svgNamespace: "http://www.w3.org/2000/svg",
		arxNamespace: "http://arxos.io/svg",
	}
}

// RenderRecognizedSymbols renders recognized symbols into SVG-BIM
func (r *SymbolRenderer) RenderRecognizedSymbols(svgContent string, recognizedSymbols []map[string]interface{},
	buildingID string, floorLabel string) RenderResult {

	result := RenderResult{
		SVG:             svgContent,
		RenderedSymbols: []RenderedSymbol{},
		TotalRecognized: len(recognizedSymbols),
		TotalRendered:   0,
		BuildingID:      buildingID,
		FloorLabel:      floorLabel,
	}

	// Parse SVG content
	svgData, err := r.parseSVG(svgContent)
	if err != nil {
		result.Error = fmt.Sprintf("Failed to parse SVG: %v", err)
		return result
	}

	// Find or create arx-objects group
	arxGroup := r.findOrCreateArxGroup(svgData)

	// Render each recognized symbol
	for _, symbolInfo := range recognizedSymbols {
		confidence, _ := symbolInfo["confidence"].(float64)

		// Only render high-confidence matches
		if confidence >= 0.5 {
			renderedSymbol := r.renderSingleSymbol(arxGroup, symbolInfo, buildingID, floorLabel)
			if renderedSymbol != nil {
				result.RenderedSymbols = append(result.RenderedSymbols, *renderedSymbol)
				result.TotalRendered++
			}
		}
	}

	// Convert back to string
	result.SVG = r.svgToString(svgData)

	return result
}

// UpdateSymbolPosition updates the position of an existing symbol
func (r *SymbolRenderer) UpdateSymbolPosition(svgContent string, objectID string, newPosition Point) string {
	svgData, err := r.parseSVG(svgContent)
	if err != nil {
		return svgContent
	}

	// Find the symbol element
	symbolElem := r.findSymbolElement(svgData, objectID)
	if symbolElem != nil {
		// Update transform
		symbolElem.Attributes["transform"] = fmt.Sprintf("translate(%.2f,%.2f)", newPosition.X, newPosition.Y)

		// Update metadata
		symbolElem.Attributes["data-updated-at"] = time.Now().Format(time.RFC3339)
	}

	return r.svgToString(svgData)
}

// RemoveSymbol removes a symbol from the SVG
func (r *SymbolRenderer) RemoveSymbol(svgContent string, objectID string) string {
	svgData, err := r.parseSVG(svgContent)
	if err != nil {
		return svgContent
	}

	// Find and remove the symbol element
	r.removeSymbolElement(svgData, objectID)

	return r.svgToString(svgData)
}

// GetRenderedSymbols extracts metadata from rendered symbols in SVG
func (r *SymbolRenderer) GetRenderedSymbols(svgContent string) []RenderedSymbol {
	var symbols []RenderedSymbol

	svgData, err := r.parseSVG(svgContent)
	if err != nil {
		return symbols
	}

	// Find all arx-symbol groups
	arxGroups := r.findArxSymbolGroups(svgData)
	for _, group := range arxGroups {
		symbol := r.extractSymbolMetadata(group)
		if symbol != nil {
			symbols = append(symbols, *symbol)
		}
	}

	return symbols
}

// Helper methods

func (r *SymbolRenderer) parseSVG(svgContent string) (*SVGData, error) {
	// This is a simplified SVG parser
	// In a real implementation, you would use a proper XML/SVG parser

	svgData := &SVGData{
		Elements: []SVGElement{},
	}

	// Basic parsing - extract groups and elements
	// This is a placeholder implementation
	// In practice, you'd use an XML parser like encoding/xml

	return svgData, nil
}

func (r *SymbolRenderer) findOrCreateArxGroup(svgData *SVGData) *SVGElement {
	// Look for existing arx-objects group
	for i := range svgData.Elements {
		if svgData.Elements[i].Tag == "g" && svgData.Elements[i].Attributes["id"] == "arx-objects" {
			return &svgData.Elements[i]
		}
	}

	// Create new arx-objects group
	arxGroup := SVGElement{
		Tag: "g",
		Attributes: map[string]string{
			"id":    "arx-objects",
			"class": "arx-dynamic-objects",
		},
		Children: []SVGElement{},
	}

	svgData.Elements = append(svgData.Elements, arxGroup)
	return &svgData.Elements[len(svgData.Elements)-1]
}

func (r *SymbolRenderer) renderSingleSymbol(arxGroup *SVGElement, symbolInfo map[string]interface{},
	buildingID string, floorLabel string) *RenderedSymbol {

	symbolID, _ := symbolInfo["symbol_id"].(string)
	symbolData, _ := symbolInfo["symbol_data"].(map[string]interface{})
	confidence, _ := symbolInfo["confidence"].(float64)
	matchType, _ := symbolInfo["match_type"].(string)

	// Generate unique object ID
	objectID := fmt.Sprintf("%s_%s", symbolID, uuid.New().String()[:8])

	// Determine position
	position := r.determineSymbolPosition(symbolInfo, arxGroup)

	// Create the symbol element
	symbolElement := r.createSymbolElement(symbolID, symbolData, objectID, position, confidence, matchType)

	// Add to arx-group
	arxGroup.Children = append(arxGroup.Children, *symbolElement)

	// Create metadata
	metadata := RenderedSymbol{
		ObjectID:    objectID,
		SymbolID:    symbolID,
		SymbolName:  r.getStringValue(symbolData, "display_name", symbolID),
		System:      r.getStringValue(symbolData, "system", "unknown"),
		Category:    r.getStringValue(symbolData, "category", ""),
		Position:    position,
		Confidence:  confidence,
		MatchType:   matchType,
		RenderedAt:  time.Now(),
		BuildingID:  buildingID,
		FloorLabel:  floorLabel,
		Properties:  r.getMapValue(symbolData, "properties"),
		Connections: r.getConnections(symbolData),
		Tags:        r.getStringSlice(symbolData, "tags"),
	}

	return &metadata
}

func (r *SymbolRenderer) determineSymbolPosition(symbolInfo map[string]interface{}, arxGroup *SVGElement) Point {
	// If we have a position from SVG text recognition, use it
	if pos, exists := symbolInfo["position"]; exists {
		if posMap, ok := pos.(map[string]interface{}); ok {
			x, _ := posMap["x"].(float64)
			y, _ := posMap["y"].(float64)
			return Point{X: x, Y: y}
		}
	}

	// Calculate position based on existing symbols
	return r.calculateAutoPosition(arxGroup, symbolInfo)
}

func (r *SymbolRenderer) calculateAutoPosition(arxGroup *SVGElement, symbolInfo map[string]interface{}) Point {
	// Get existing symbol positions
	existingPositions := r.getExistingPositions(arxGroup)

	// Calculate new position
	if len(existingPositions) > 0 {
		// Find a position that doesn't overlap
		baseX, baseY := existingPositions[len(existingPositions)-1].X, existingPositions[len(existingPositions)-1].Y
		offset := 50.0

		// Try positions in a grid pattern
		for i := 0; i < 10; i++ {
			x := baseX + float64(i%3)*offset
			y := baseY + float64(i/3)*offset

			// Check if position is far enough from existing symbols
			if !r.isPositionOverlapping(x, y, existingPositions) {
				return Point{X: x, Y: y}
			}
		}

		// If no good position found, just offset from last
		lastPos := existingPositions[len(existingPositions)-1]
		return Point{X: lastPos.X + offset, Y: lastPos.Y + offset}
	}

	// First symbol - place at origin
	return Point{X: 100, Y: 100}
}

func (r *SymbolRenderer) getExistingPositions(arxGroup *SVGElement) []Point {
	var positions []Point

	for _, child := range arxGroup.Children {
		if transform, exists := child.Attributes["transform"]; exists {
			if pos := r.extractPositionFromTransform(transform); pos != nil {
				positions = append(positions, *pos)
			}
		}
	}

	return positions
}

func (r *SymbolRenderer) extractPositionFromTransform(transform string) *Point {
	// Parse translate(x,y) from transform
	re := regexp.MustCompile(`translate\(([^,]+),([^)]+)\)`)
	matches := re.FindStringSubmatch(transform)
	if len(matches) == 3 {
		if x, err := r.parseFloat(matches[1]); err == nil {
			if y, err := r.parseFloat(matches[2]); err == nil {
				return &Point{X: x, Y: y}
			}
		}
	}
	return nil
}

func (r *SymbolRenderer) parseFloat(s string) (float64, error) {
	// Simple float parsing - in practice you'd use strconv.ParseFloat
	return 0, fmt.Errorf("not implemented")
}

func (r *SymbolRenderer) isPositionOverlapping(x, y float64, existingPositions []Point) bool {
	for _, pos := range existingPositions {
		distance := math.Sqrt(math.Pow(x-pos.X, 2) + math.Pow(y-pos.Y, 2))
		if distance < 30 {
			return true
		}
	}
	return false
}

func (r *SymbolRenderer) createSymbolElement(symbolID string, symbolData map[string]interface{},
	objectID string, position Point, confidence float64, matchType string) *SVGElement {

	// Get the symbol's SVG content
	symbolSVG := r.getStringValue(symbolData, "svg", "")

	// Create the main group element
	groupAttrs := map[string]string{
		"id":               objectID,
		"class":            fmt.Sprintf("arx-symbol arx-%s", symbolID),
		"transform":        fmt.Sprintf("translate(%.2f,%.2f)", position.X, position.Y),
		"data-symbol-id":   symbolID,
		"data-symbol-name": r.getStringValue(symbolData, "display_name", symbolID),
		"data-system":      r.getStringValue(symbolData, "system", "unknown"),
		"data-category":    r.getStringValue(symbolData, "category", ""),
		"data-confidence":  fmt.Sprintf("%.2f", confidence),
		"data-match-type":  matchType,
		"data-rendered-at": time.Now().Format(time.RFC3339),
	}

	group := &SVGElement{
		Tag:        "g",
		Attributes: groupAttrs,
		Children:   []SVGElement{},
	}

	// Add the symbol SVG content
	if symbolSVG != "" {
		if symbolRoot, err := r.parseSVG(symbolSVG); err == nil {
			// Copy all child elements
			group.Children = append(group.Children, symbolRoot.Elements...)
		} else {
			// Fallback: add as text
			textElem := SVGElement{
				Tag: "text",
				Attributes: map[string]string{
					"x":           "0",
					"y":           "0",
					"font-size":   "12",
					"text-anchor": "middle",
				},
				Text: r.getStringValue(symbolData, "display_name", symbolID),
			}
			group.Children = append(group.Children, textElem)
		}
	}

	// Add confidence indicator
	if confidence < 0.8 {
		confidenceColor := "#ff6b6b"
		if confidence >= 0.6 {
			confidenceColor = "#ffd93d"
		}

		confidenceIndicator := SVGElement{
			Tag: "circle",
			Attributes: map[string]string{
				"cx":               "0",
				"cy":               "0",
				"r":                "15",
				"fill":             "none",
				"stroke":           confidenceColor,
				"stroke-width":     "2",
				"stroke-dasharray": "5,5",
				"opacity":          "0.7",
			},
		}
		group.Children = append(group.Children, confidenceIndicator)
	}

	// Add metadata text
	metadataText := SVGElement{
		Tag: "text",
		Attributes: map[string]string{
			"x":           "0",
			"y":           "25",
			"font-size":   "8",
			"text-anchor": "middle",
			"fill":        "#666",
			"class":       "arx-metadata",
		},
		Text: fmt.Sprintf("%s (%.1f)", r.getStringValue(symbolData, "display_name", symbolID), confidence),
	}
	group.Children = append(group.Children, metadataText)

	return group
}

func (r *SymbolRenderer) findSymbolElement(svgData *SVGData, objectID string) *SVGElement {
	// Recursively search for element with matching ID
	return r.findElementByID(svgData.Elements, objectID)
}

func (r *SymbolRenderer) findElementByID(elements []SVGElement, objectID string) *SVGElement {
	for i := range elements {
		if elements[i].Attributes["id"] == objectID {
			return &elements[i]
		}
		if len(elements[i].Children) > 0 {
			if found := r.findElementByID(elements[i].Children, objectID); found != nil {
				return found
			}
		}
	}
	return nil
}

func (r *SymbolRenderer) removeSymbolElement(svgData *SVGData, objectID string) {
	// Remove element by ID from the SVG data structure
	r.removeElementByID(&svgData.Elements, objectID)
}

func (r *SymbolRenderer) removeElementByID(elements *[]SVGElement, objectID string) {
	for i := len(*elements) - 1; i >= 0; i-- {
		if (*elements)[i].Attributes["id"] == objectID {
			// Remove element
			*elements = append((*elements)[:i], (*elements)[i+1:]...)
			return
		}
		if len((*elements)[i].Children) > 0 {
			r.removeElementByID(&(*elements)[i].Children, objectID)
		}
	}
}

func (r *SymbolRenderer) findArxSymbolGroups(svgData *SVGData) []*SVGElement {
	var groups []*SVGElement
	r.findGroupsByClass(svgData.Elements, "arx-symbol", &groups)
	return groups
}

func (r *SymbolRenderer) findGroupsByClass(elements []SVGElement, className string, groups *[]*SVGElement) {
	for i := range elements {
		if elements[i].Tag == "g" && strings.Contains(elements[i].Attributes["class"], className) {
			*groups = append(*groups, &elements[i])
		}
		if len(elements[i].Children) > 0 {
			r.findGroupsByClass(elements[i].Children, className, groups)
		}
	}
}

func (r *SymbolRenderer) extractSymbolMetadata(group *SVGElement) *RenderedSymbol {
	symbolData := RenderedSymbol{
		ObjectID:   group.Attributes["id"],
		SymbolID:   group.Attributes["data-symbol-id"],
		SymbolName: group.Attributes["data-symbol-name"],
		System:     group.Attributes["data-system"],
		Category:   group.Attributes["data-category"],
		MatchType:  group.Attributes["data-match-type"],
	}

	// Parse confidence
	if confidenceStr := group.Attributes["data-confidence"]; confidenceStr != "" {
		if confidence, err := r.parseFloat(confidenceStr); err == nil {
			symbolData.Confidence = confidence
		}
	}

	// Parse position from transform
	if transform := group.Attributes["transform"]; transform != "" {
		if pos := r.extractPositionFromTransform(transform); pos != nil {
			symbolData.Position = *pos
		}
	}

	// Parse rendered_at
	if renderedAtStr := group.Attributes["data-rendered-at"]; renderedAtStr != "" {
		if renderedAt, err := time.Parse(time.RFC3339, renderedAtStr); err == nil {
			symbolData.RenderedAt = renderedAt
		}
	}

	return &symbolData
}

func (r *SymbolRenderer) svgToString(svgData *SVGData) string {
	// Convert SVG data structure back to string
	// This is a placeholder implementation
	return "<svg>...</svg>"
}

// Helper methods for data extraction

func (r *SymbolRenderer) getStringValue(data map[string]interface{}, key string, defaultValue string) string {
	if value, exists := data[key]; exists {
		if str, ok := value.(string); ok {
			return str
		}
	}
	return defaultValue
}

func (r *SymbolRenderer) getMapValue(data map[string]interface{}, key string) map[string]interface{} {
	if value, exists := data[key]; exists {
		if mapValue, ok := value.(map[string]interface{}); ok {
			return mapValue
		}
	}
	return make(map[string]interface{})
}

func (r *SymbolRenderer) getStringSlice(data map[string]interface{}, key string) []string {
	if value, exists := data[key]; exists {
		if slice, ok := value.([]interface{}); ok {
			var result []string
			for _, item := range slice {
				if str, ok := item.(string); ok {
					result = append(result, str)
				}
			}
			return result
		}
	}
	return []string{}
}

func (r *SymbolRenderer) getConnections(data map[string]interface{}) []Connection {
	if value, exists := data["connections"]; exists {
		if slice, ok := value.([]interface{}); ok {
			var connections []Connection
			for _, item := range slice {
				if connMap, ok := item.(map[string]interface{}); ok {
					conn := Connection{
						ID:   r.getStringValue(connMap, "id", ""),
						Type: r.getStringValue(connMap, "type", ""),
					}
					if posData, exists := connMap["position"]; exists {
						if posMap, ok := posData.(map[string]interface{}); ok {
							x, _ := posMap["x"].(float64)
							y, _ := posMap["y"].(float64)
							conn.Position = Point{X: x, Y: y}
						}
					}
					connections = append(connections, conn)
				}
			}
			return connections
		}
	}
	return []Connection{}
}

// SVGData represents the structure of SVG data
type SVGData struct {
	Elements []SVGElement
}

// SVGElement represents an SVG element
type SVGElement struct {
	Tag        string            `json:"tag"`
	Attributes map[string]string `json:"attributes"`
	Text       string            `json:"text,omitempty"`
	Children   []SVGElement      `json:"children,omitempty"`
}
